protocol/msgpack/base_iterator.go

```diff
@@ -61,15 +66,40 @@ func (it *baseIterator) err() error       { return it.decodeErr }
 func (it *baseIterator) setErr(err error) { it.decodeErr = err }
 
 func (it *baseIterator) decodePolicy() policy.Policy {
-	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(policyType)
+	numActualFields := it.decodeNumObjectFields()
+	policyType := it.decodeObjectType()
+
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
+		policyType,
+		numActualFields,
+	)
 	if !ok {
 		return policy.EmptyPolicy
 	}
-	resolution := it.decodeResolution()
-	retention := it.decodeRetention()
-	p := policy.NewPolicy(resolution.Window, resolution.Precision, time.Duration(retention))
-	it.skip(numActualFields - numExpectedFields)
-	return p
+
+	switch policyType {
+	case rawPolicyType:
+		resolution := it.decodeResolution()
+		retention := it.decodeRetention()
+		p := policy.NewPolicy(resolution.Window, resolution.Precision, time.Duration(retention))
+		it.skip(numActualFields - numExpectedFields)
+		return p
+	case compressedPolicyType:
+		id := it.decodeVarint()
```

Comment:
The decompressor is set by the options passed into the iterator. When you create a new set of options for the iterator the decompressor in the options is set to the no-op decompressor. So for the decompressor to be nil you would have to explicitly set it to nil. My thinking is that we can check it once when we construct the decompressor and if its nil we can set it to no-op so we never have to check that its nil when we run the iterator. I don't feel strongly either way though so I'm fine with making the default nil and then just doing a check before we call it. The only downside is that its an extra operation when its not nil, but thats a trivial cost.

As for adding an enabled flag, that was my first approach, but I decided not to include one since it wouldn't effect the actual operation of the iterator since if we encounter an id we don't recognize we wont be able to decompress it either way. It may make debugging slightly easier so the error will indicate whether decompression is even enabled or not.